client.go

```diff
@@ -1830,3 +1836,33 @@ func flags(f int) uint32 {
 	}
 	return out
 }
+
+// posixPerm converts Go permission bits to POSIX permission bits,
+// masking off file type bits.
+//
+// This differs from fromFileMode in that it preserves the POSIX versions of
+// setuid, setgid and sticky in m, because we've historically supported those
+// bits.
+func posixPerm(m os.FileMode) uint32 {
```

Comment:
I think we could change `fromFileMode` to incorporate this “support `S_ISUID` and `os.ModeSetuid`” in order to get the baseline permissions at the start, then it could go on to the next step of adding the additional `S_IFMT` values in.

It looks like there are bugs in `fromFileMode` anyways, because it could potentially OR in various `S_IFMT` values potentially all together, when only one of them should ever be valid at any one time. (i.e. since they all share the same bit mask, accidentally composing `S_IFREG | S_IFDIR` gets you `S_IFLNK` _not_ what was intended. In this case, it should probably return an `error`.)